server/server/src/billing/billing_service.rs

```diff
@@ -73,6 +79,67 @@ fn release_subscription_profile<T: Debug>(
     })?)
 }
 
+pub async fn upgrade_account_app_store(
+    context: RequestContext<'_, UpgradeAccountAppStoreRequest>,
+) -> Result<UpgradeAccountAppStoreResponse, ServerError<UpgradeAccountAppStoreError>> {
+    let (request, server_state) = (&context.request, context.server_state);
+
+    let mut account = lock_subscription_profile(server_state, &context.public_key)?;
+
+    if account.billing_info.data_cap() == PREMIUM_TIER_USAGE_SIZE {
+        return Err(ClientError(UpgradeAccountAppStoreError::AlreadyPremium));
+    }
+
+    debug!("Upgrading the account of a user through app store billing");
+
+    if let Some(owner) = server_state
+        .index_db
+        .app_store_ids
+        .get(&request.app_account_token)?
+    {
+        if let Some(other_account) = server_state.index_db.accounts.get(&owner)? {
+            if other_account.billing_info.billing_platform.is_some() {
+                return Err(ClientError(UpgradeAccountAppStoreError::AppStoreAccountAlreadyLinked));
+            }
+        }
+    }
+
+    let expires = verify_details(
```

Comment:
Do we need to check this `expires` for validity?

Reply:
We do not since in `verify_details`, the subscription status endpoint we make a request to tells us if the subscription has expired or not.